go/arrow/avro/schema.go

```diff
@@ -0,0 +1,407 @@
+	case "null":
+		n.arrowField = arrow.Field{Name: n.name, Type: arrow.BinaryTypes.Binary, Nullable: true}
+	}
+}
+
+// iterate record Fields()
+func iterateFields(n *schemaNode) {
+	for _, f := range n.schema.(avro.PropertySchema).(*avro.RecordSchema).Fields() {
+		switch ft := f.Type().(type) {
+		// Avro "array" field type
+		case *avro.ArraySchema:
+			schemaCache.Add(f.Name(), ft.Items())
+			// logical items type
+			c := n.newChild(f.Name(), ft.Items())
+			if isLogicalSchemaType(ft.Items()) {
+				avroLogicalToArrowField(c)
+			} else {
+				arrowSchemafromAvro(c)
+			}
+			c.arrowField = arrow.Field{Name: c.name, Type: arrow.ListOf(c.arrowField.Type), Metadata: c.arrowField.Metadata, Nullable: true}
+		// Avro "enum" field type = Arrow dictionary type
+		case *avro.EnumSchema:
+			schemaCache.Add(f.Name(), f.Type())
+			c := n.newChild(f.Name(), f.Type())
+			symbols := make(map[string]string)
+			for index, symbol := range ft.Symbols() {
+				k := strconv.FormatInt(int64(index), 10)
+				symbols[k] = symbol
+			}
+			var dt arrow.DictionaryType = arrow.DictionaryType{IndexType: arrow.PrimitiveTypes.Uint64, ValueType: arrow.BinaryTypes.String, Ordered: false}
+			sl := len(symbols)
+			switch {
+			case sl <= math.MaxUint8:
+				dt.IndexType = arrow.PrimitiveTypes.Uint8
+			case sl > math.MaxUint8 && sl <= math.MaxUint16:
+				dt.IndexType = arrow.PrimitiveTypes.Uint16
+			case sl > math.MaxUint16 && sl <= math.MaxUint32:
+				dt.IndexType = arrow.PrimitiveTypes.Uint32
+			}
+			c.arrowField = arrow.Field{Name: f.Name(), Type: &dt, Nullable: true, Metadata: arrow.MetadataFrom(symbols)}
+		// Avro "fixed" field type = Arrow FixedSize Primitive BinaryType
+		case *avro.FixedSchema:
+			schemaCache.Add(f.Name(), f.Type())
+			if isLogicalSchemaType(f.Type()) {
+				c := n.newChild(f.Name(), f.Type())
+				avroLogicalToArrowField(c)
+			} else {
+				c := n.newChild(f.Name(), f.Type())
+				arrowSchemafromAvro(c)
+			}
+		case *avro.RecordSchema:
+			schemaCache.Add(f.Name(), f.Type())
+			c := n.newChild(f.Name(), f.Type())
+			iterateFields(c)
+			// Avro "map" field type - KVP with value of one type - keys are strings
+		case *avro.MapSchema:
+			schemaCache.Add(f.Name(), ft.Values())
+			c := n.newChild(f.Name(), ft.Values())
+			arrowSchemafromAvro(c)
+			if ft.Values().Type() == "union" {
+				c.arrowField = arrow.Field{Name: f.Name(), Type: arrow.MapOf(arrow.BinaryTypes.String, arrow.BinaryTypes.String), Metadata: c.arrowField.Metadata, Nullable: true}
+			} else {
+				c.arrowField = arrow.Field{Name: f.Name(), Type: arrow.MapOf(arrow.BinaryTypes.String, c.arrowField.Type), Metadata: c.arrowField.Metadata, Nullable: true}
+			}
+		case *avro.UnionSchema:
+			if ft.Nullable() {
+				if len(ft.Types()) > 1 {
+					schemaCache.Add(f.Name(), ft.Types()[1])
+					c := n.newChild(f.Name(), ft.Types()[1])
+					c.union = true
+					arrowSchemafromAvro(c)
+				}
+			}
+		default:
+			schemaCache.Add(f.Name(), f.Type())
```

Comment:
could there end up being multiple *different* schemas with the same name? Would that be a problem for this cache?